docs/api_reference/src/quantities.tex

```diff
@@ -109,158 +599,4907 @@
 
-\begin{codeblock}
-namespace mp_units {
+  // comparison
+  template<std::size_t N2, std::size_t M2>
+  friend constexpr auto operator<=>(const symbol_text& lhs,
+                                    const symbol_text<N2, M2>& rhs) noexcept;
+  template<std::size_t N2, std::size_t M2>
+  friend constexpr bool operator==(const symbol_text& lhs,
+                                   const symbol_text<N2, M2>& rhs) noexcept;
+};
+
+// deduction guides
+
+symbol_text(char) -> symbol_text<1, 1>;
+
+template<std::size_t N>
+symbol_text(const char (&)[N]) -> symbol_text<N - 1, N - 1>;
+
+template<std::size_t N>
+symbol_text(const std::fixed_string<N>&) -> symbol_text<N, N>;
+
+template<std::size_t N, std::size_t M>
+symbol_text(const char8_t (&)[N], const char (&)[M]) -> symbol_text<N - 1, M - 1>;
 
-export template<@\unspec@>
-class quantity_point { @\unspec@ };
+template<std::size_t N, std::size_t M>
+symbol_text(const std::fixed_u8string<N>&, const std::fixed_string<M>&) -> symbol_text<N, M>;
 
 }
 \end{codeblock}
 
-A \defnadj{quantity point}{type} is a specialization of \tcode{quantity_point}.
-Let \tcode{\placeholder{Q}} be a quantity point type.
-\tcode{\placeholdernc{Q}::point_origin} represents
-the origin point of a position vector\irefiev{102-03-15}.
+\pnum
+\tcode{symbol_text} represents a symbol text.
+\exposid{utf8} stores its UTF-8 representation, and
+\exposid{portable} stores its portable representation.
+\tcode{symbol_text<N, M>} is a structural type\irefcppx{temp.param}{term.structural.type}.
+
+\pnum
+In the descriptions that follow,
+it is a \Fundescx{Precondition} that
 \begin{itemize}
 \item
-If \tcode{Rep} is a scalar,
-\tcode{\placeholder{Q}} represents the scalar quantity\irefiev{102-02-19}
-of a position vector.
+values of \tcode{char} are in the basic literal character set\irefcpp{lex.charset}, and
 \item
-If \tcode{Rep} is a vector,
-\tcode{\placeholder{Q}} represents a position vector.
-% FIXME What if `Rep` is a tensor?
+for a parameter of the form \tcode{const \placeholdernc{CharT} (\&\placeholdernc{txt})[\placeholder{M}]},
+\tcode{(\placeholdernc{txt}[\placeholdernc{M} - 1] == \placeholdernc{CharT}())} is \tcode{true}.
 \end{itemize}
 
-\rSec1[qty.compat]{Compatibility}
+\indexlibraryctor{symbol_text}
+\begin{itemdecl}
+constexpr explicit(false) symbol_text(char portable);
+consteval explicit(false) symbol_text(const char (&portable)[N + 1]);
+constexpr explicit(false) symbol_text(const std::fixed_string<N>& portable);
+consteval symbol_text(const char8_t (&utf8)[N + 1], const char (&portable)[M + 1]);
+constexpr symbol_text(const std::fixed_u8string<N>& utf8, const std::fixed_string<M>& portable);
+\end{itemdecl}
 
-\rSec1[qty.one]{Dimension one}
+\begin{itemdescr}
+\pnum
+Let \tcode{lhs} be \tcode{utf8} for the last two constructors,
+and otherwise:
+\begin{codeblock}
+std::bit_cast<std::fixed_u8string<N>>(std::basic_fixed_string(portable))
```

Comment:
Searching the C++ Standard library for "trivial",
it's obvious that we can't continue using `std::bit_cast` here
unless `std::basic_fixed_string` has wording like that of commit 2c2377e63c0f028f3dd932a9003d76f1be50cf1b
or is specified as a trivially copyable type.
Or maybe I should just use `std::ranges::to` here.

Reply:
Maybe SG16 has an opinion on converting between `char8_t` and `char` arrays.

Reply:
Oh, I think I know.
The `bit_cast` isn't guaranteed to work.
The values of `char` might not be UTF-8 compatible.

Reply:
In fact, we'd need a dedicated algorithm
to convert from UTF-8 to whatever-it-was.
Is this implementable in the C++ Standard yet?

Reply:
SG16 said to do exactly the same as `print` does, and I think they use `bit_cast` there.

Reply:
We discussed this a bit with Victor Zverovich at the last ISO meeting. See: febdaec85e4148362c2c8a06ba204818893ce2a6.

Reply:
What does that mean for the problem here?

Reply:
If `text_encoding` for `char` is UTF-8, then it is allowed to convert from UTF-8 to `char` with just `static_cast`.

I do not remember the details, but the `bit_cast` for `fixed_string` was also probably recommended by SG16.

Reply:
OK, that makes sense.

But a string literal like `"pi"`,
which is used to initialize a `fixed_string`,
is still not required to be UTF-8 or ASCII.

So I think all initializers for `fixed_string` should use a `u8""` string literal,
even if we still statically check them to only include `char`s in the basic literal character set.

Reply:
> So I think all initializers for `fixed_string` should use a `u8""` string literal,
> even if we still statically check them to only include `char`s in the basic literal character set.

Well, I'm not too sure about that.
Is it OK for `fixed_string` to not store ASCII or UTF-8?
Do the print functions working with `char`
know that they are the character set used by the ordinary string literals and not ASCII or UTF-8?
I suppose they do, but I don't actually know.

Reply:
Whatever, what printing `char` does is irrelevant.
The problem here is that `bit_cast`ing from an ordinary string literal whose character set is not ASCII or UTF-8
puts garbage in the _`utf8`_ member that isn't actually UTF-8.
So when you go to print this, or use it to sort some list, you'll get non-portable results.

Reply:
I am not sure if I fully understand what you mean. We call `to_u8string` with the arguments always being characters from the basic literal character set. Maybe, as of now, in mp-units, we do check it after the `bit_cast` is done, which is wrong, but it is meant to be replaced with proper contracts and `Requires` sections.

Reply:
`to_u8string` isn't in these reference documentation yet.

Basically, the bit pattern of `'a'` isn't required to be the same as that of `u8'a'`.
The same applies to all other characters in the basic literal character set.
Maybe if we could such an implementation in Compiler Explorer,
we'd be able to verify that using `bit_cast` to initialize `utf8` from `portable`
would give the wrong answer when printing it.